src/Passes/README.md

```diff
@@ -0,0 +1,322 @@
+}
+```
+
+into following IR:
+
+```
+define double @test(double %x) {
+entry:
+        %addtmp = fadd double 3.000000e+00, %x
+        %addtmp1 = fadd double %x, 3.000000e+00
+        %multmp = fmul double %addtmp, %addtmp1
+        ret double %multmp
+}
+```
+
+This code is obviously inefficient as we could get rid of one operation by rewritting the code to:
+
+```c
+double test(double x) {
+    double y = 3+x;
+    return y * y;
+}
+```
+
+One purpose of LLVM passes is to allow automatic transformation from the above IR to the IR:
+
+```
+define double @test(double %x) {
+entry:
+        %addtmp = fadd double %x, 3.000000e+00
+        %multmp = fmul double %addtmp, %addtmp
+        ret double %multmp
+}
+```
+
+**Example 2: Analytics**. Another example of useful passes are those generating and collecting statistics about the program. For instance, one analytics program
+makes sense for classical programs is to count instructions used to implement functions. Take the C program:
+
+```c
+int foo(int x)
+{
+  return x;
+}
+
+void bar(int x, int y)
+{
+  foo(x + y);
+}
+
+int main()
+{
+  foo(2);
+  bar(3, 2);
+
+  return 0;
+}
+```
+
+which produces follow IR (without optimisation):
+
+```language
+define dso_local i32 @foo(i32 %0) #0 {
+  %2 = alloca i32, align 4
+  store i32 %0, i32* %2, align 4
+  %3 = load i32, i32* %2, align 4
+  ret i32 %3
+}
+
+define dso_local void @bar(i32 %0, i32 %1) #0 {
+  %3 = alloca i32, align 4
+  %4 = alloca i32, align 4
+  store i32 %0, i32* %3, align 4
+  store i32 %1, i32* %4, align 4
+  %5 = load i32, i32* %3, align 4
+  %6 = load i32, i32* %4, align 4
+  %7 = add nsw i32 %5, %6
+  %8 = call i32 @foo(i32 %7)
+  ret void
+}
+
+define dso_local i32 @main() #0 {
+  %1 = alloca i32, align 4
+  store i32 0, i32* %1, align 4
+  %2 = call i32 @foo(i32 2)
+  call void @bar(i32 3, i32 2)
+  ret i32 0
+}
+```
+
+A stat pass for this code, would collect following statisics:
+
+```text
+Stats for 'foo'
+===========================
+Opcode          # Used
+---------------------------
+load            1
+ret             1
+alloca          1
+store           1
+---------------------------
+
+Stats for 'bar'
+===========================
+Opcode          # Used
+---------------------------
+load            2
+add             1
+ret             1
+alloca          2
+store           2
+call            1
+---------------------------
+
+Stats for 'main'
+===========================
+Opcode          # Used
+---------------------------
+ret             1
+alloca          1
+store           1
+call            2
+---------------------------
+```
+
+**Example 3: Code validation**. A third use case is code validation. For example, one could write a pass to check whether bounds are exceeded on static arrays [2].
+Note that this is a non-standard usecase as such analysis is usually made using the AST rather than at the IR level.
+
+**References**
+[1] https://github.com/banach-space/llvm-tutor#analysis-vs-transformation-pass
```

Comment:
FYI: I see a reference to [2], but I don't see any reference to [1].

Reply:
Fixed.